### ISSUE TITLE
Heuristics for quality control

### ISSUE THREAD

**Issue by constantinpape** [hci-unihd/batchlib#64]
As we have discussed already, it would be good to have some additional quality control based on 
heuristics computed from the features we extract from the cell segmentation + image channels.
I have adapted the code in #51 now so that this is easy to integrate.
Now we only need to come up with (and evaluate!) good heuristics.
Some initial ideas:
- number of cells
- number of infected vs control cells
- cell size distribution
- negative ratios

cc @wolny @imagirom @tischi

**Comment by tischi** [comment]
I think also a cell based QC 
- number of pixels of a cell (reject too few or too many)

**Comment by constantinpape** [comment]
> I think also a cell based QC
> 
>     * number of pixels of a cell (reject too few or too many)

In general I agree that this makes sense.

Right now I am not so sure that we need it though; I think that due to the very good seeds we shouldn't really have a problem with segmented cells that are completely off.
(The segmentation might of course still be incorrect, but I don't think we can easily catch with size based heuristics, which btw also applies to the `cell size distribution` idea I mentioned above).

Or have you seen some examples where we could find outliers with a simple size threshold when you looked at the data, @tischi ?

**Comment by tischi** [comment]
I have not look yet. Is this plate a good one to inspect?
"/g/kreshuk/data/covid/sandbox/h5-tables2/titration_plate_20200403_154849"

**Comment by tischi** [comment]
@constantinpape 
I posted a couple of issues. Since our analysis is median based I do not think it has high priority but I think just adding one threshold for maximal cell size would be easy and remove couple of wrong segmentations. One could ask @metavibor to provide a biologically justified maximal cell size.

**Comment by constantinpape** [comment]
> @constantinpape
> I posted a couple of issues. Since our analysis is median based I do not think it has high priority but I think just adding one threshold for maximal cell size would be easy and remove couple of wrong segmentations. One could ask @metavibor to provide a biologically justified maximal cell size.

Thanks, I will have a look at those issues later.

I already discussed with @imagirom how we will handle this in the code:
One of the cell tables will have a column `cell_is_outlier` (or so).
This will be initialized all zeros by `InstanceFeatureExtraction`.

Then we can write a `CellBasedQC` job that goes over all the extracted features and can assign outliers based on the heuristics we come up with. In `CellLevelAnalysis`, the cells that were marked as outliers will then be excluded .

**Comment by tischi** [comment]
> One of the cell tables will have a column `cell_is_outlier`

Maybe be try to be consistent here: for the images it currently is called `marked_as_outlier`. For me, simply `outlier` would do.

**Comment by constantinpape** [comment]
> Maybe be try to be consistent here: for the images it currently is called `marked_as_outlier`. For me, simply `outlier` would do.

Yes good point, we should be consistent with the column names here.
I also don't really mind which name we use.

**Comment by constantinpape** [comment]
Ok, I will implement the code for the QC later (probably not the actual heuristics yet):
There will be a file `cell_based_qc.py` with the following jobs:
- `CellLevelQC`: heuristics based QC per cell; write results to a table per image
- `ImageLevelQC`: heuristic and manual (if available) QC; gets the `outlier_predicate`; writes result to a table in `PLATENAME_tables.hdf5`
- `WellLevelQC`: same as image level but for wells

The `CellLevelAnalysis` then looks for these tables and if they exists use the results in there to exclude outliers from the analysis. (No need to pass it `outlier_predicate` any more)

**Comment by metavibor** [comment]
I looked at cell sizes, the largest cell size that I could measure in a subset of images was 8500 pixels, let's but an upper threshold for the cell size on 10000 pixels

**Comment by constantinpape** [comment]
Ok, I have implemented the mechanism for image level qc and cell level qc in #77 now.
Will have a look at the issues @tischi made in the other repo later + use the size threshold @metavibor provided to implement a first heuristic later.

**Comment by constantinpape** [comment]
@tischi 
I had a look at all the issues you posted in the other repo now.
So I think for now you would also just propose to use a size threshold, right?

**Comment by tischi** [comment]
Right! As said, overall it looks very good!

**Comment by constantinpape** [comment]
I am implementing the heuristic qc now in #85 and I am almost done with it for the per cell level.
One more thing about this though:
In addition to a maximum size, we should also have a minimum size. The implementation is trivial.
@tischi @metavibor can you estimate a good value for this from the data?

**Comment by metavibor** [comment]
@constantinpape @tischi I did some measurements and 1000 seems like a good minimal threshold, I have measured a few cells smaller than this (when they get squeezed between other cells they can get quite small) but those are difficult to segment and I don't mind if we threshold them out

**Comment by constantinpape** [comment]
I have started to implement the heuristic image level QC in #90 using the following heuristics:
- minimal number of cells
- minimal number of control cells
- negative ratios

However, I realized that all these can be computed on Well level, so I think it makes more sense to move those there.
@imagirom @tischi @metavibor 
So I have a couple of questions about this:
- Would you rather have `min_number_of_control_cells` or `min_ratio_of_control_cells`?
- We still need good values for the min numbers.
- Are there any other criteria you want on well level?
- For now we wouldn't have any heurisitic on image level. This is fine, but if you can think of something else there I can implement it.

**Comment by metavibor** [comment]
> Would you rather have min_number_of_control_cells or min_ratio_of_control_cells?
We still need good values for the min numbers.
Are there any other criteria you want on well level?
For now we wouldn't have any heurisitic on image level. This is fine, but if you can think of something else there I can implement it.

not sure about any of this. I don't have a good feeling about the data output and how many control cells are we typically dealing with. We need to look at some cases of failed analysis to figure this out.

> Would you rather have min_number_of_control_cells or min_ratio_of_control_cells?

or both. not sure what "min_ratio_of_control_cells" actually is and in which cases it indicates QC issue

> We still need good values for the min numbers.

I suggest to start with minimal number of control cells = 50. And work our way from there. If we see something failed and we got weird ratio because of the low number of control cells than we will increase it.

> Are there any other criteria you want on well level?

Maybe @tischi has some ideas based on his experience with this types of screens. I think we need to look at examples of images that failed QC control that Severina did and from them figure relevant criteria

**Comment by tischi** [comment]
I agree with @metavibor we have to look at the data first before we can decide...

I would start with having both:
- `min_number_of_control_cells` and put this to `5` for now. 
- `min_fraction_of_control_cells` and put this to `0.05` (5 %) for now.

**Comment by constantinpape** [comment]
> I agree with @metavibor we have to look at the data first before we can decide...
> 
> I would start with having both:
> 
>     * `min_number_of_control_cells` and put this to `5` for now.
> 
>     * `min_fraction_of_control_cells` and put this to `0.05` (5 %) for now.

Yes, I definitely agree, we need to derive these values from the data / looking at it.
I will start with your suggestions and then we can refine this over time.

Also let me know if you have any more thoughts on image level qc.

Thanks!

**Comment by tischi** [comment]
@constantinpape 

1. Above should be well level QCs, right?
2. On the image level we usually (a) compute a focus quality score and (b) try to detect whether there was some large debris and (c) require a `min_num_cells`, let's start with `10`. The reasoning for (c) is that usually if there are to few cells they are `not happy` and their biological response to all kind of things could change.

**Comment by constantinpape** [comment]
> * Above should be well level QCs, right?
Yes.
> 
> * On the image level we usually (a) compute a focus quality score and (b) try to detect whether there was some large debris

Ok, sounds reasonable, but we would need to figure out how to compute this here (from python).

> and (c) require a `min_num_cells`, let's start with `10`. The reasoning for (c) is that usually if there are too few cells they are "not happy" and their biological response to all kind of things could change. Also a `max_num_cells` should be considered, but again, we need to look at data. So maybe just put `max_num_cells` to `1000` for now.

Ok, makes sense I will add min and max num cells to the image qc and set it to the values you propose in the beginning.

**Comment by constantinpape** [comment]
Ok, I have implemented this in #90 now. The current settings for the parameters used in the heuristics are here:
https://github.com/hci-unihd/batchlib/blob/master/batchlib/analysis/cell_analysis_qc.py#L9-L22
What's left to do is to properly get the values for these parameters from the data.

**Comment by constantinpape** [comment]
Closing this in favor of a new issue.
